bats/tests/k8s/helm-install-rancher.bats

```diff
@@ -13,12 +13,95 @@ local_setup() {
     helm repo update
 }
 
+# Get the minimum Kubernetes version the given rancher-latest/rancher helm chart
+# does _not_ support; i.e. the chart supports < [output].
+get_chart_unsupported_version() {
+    output="" # Hack: tell shellcheck about this variable
+    local chart_version=$1
+    run helm show chart rancher-latest/rancher --version "$chart_version"
+    assert_success || return
+    run awk '/^kubeVersion:/ { $1 = ""; print }' <<<"$output"
+    assert_success || return
+    # We only support kubeVersion of form "< x.y.z"
+    assert_output --regexp '^[[:space:]]*<[[:space:]]*[^[:space:]]+$' || return
+    awk '{ print $2 }' <<<"$output"
+}
+
+# Try to determine the best chart version for the current Kubernetes version
+# (as specified in $RD_KUBERNETES_PREV_VERSION).  Saves the resulting version
+# in rancher_chart_version (and should be restored via load_var).
+determine_chart_version() {
+    if [[ -n $RD_RANCHER_IMAGE_TAG ]]; then
+        # If a version is given, check that it's compatible.
+        assert_rancher_chart_compatible "${RD_RANCHER_IMAGE_TAG#v}" || return
+        rancher_chart_version=${RD_RANCHER_IMAGE_TAG#v}
+        save_var rancher_chart_version
+        return
+    fi
+    local default_version
+    default_version=$(rancher_image_tag)
+    default_version=${default_version#v}
+    run --separate-stderr helm search repo --versions rancher-latest/rancher --output json
+    assert_success || return
+    local versions_json=$output
+    run jq_output 'map(.version).[]'
+    assert_success || return
+    run sort --version-sort <<<"$output"
+    assert_success || return
+    local versions=$output
+    local version
+    for version in $versions; do
+        if ! semver_is_valid "$version"; then
+            continue # Skip invalid / RC versions.
+        fi
+        if semver_lt "$version" "$default_version"; then
+            continue # Skip any versions older than the default version
+        fi
+        run get_chart_unsupported_version "$version"
+        assert_success || return
+        local unsupported_version=$output
+        if semver_lt "$RD_KUBERNETES_PREV_VERSION" "$unsupported_version"; then
+            # Once we find a compatible version, use it (and don't look at the
+            # rest of the chart versions).
+            trace "$(printf "Selected rancher chart version %s (wants < %s, have %s)" \
+                "$version" "$unsupported_version" "$RD_KUBERNETES_PREV_VERSION")"
+            rancher_chart_version=$version
+            save_var rancher_chart_version
+            return
+        fi
+    done
+    skip_k3s_version
+    printf "Could not find a version of rancher-latest/rancher compatible with Kubernetes %s\n" \
+        "$RD_KUBERNETES_PREV_VERSION" |
+        fail || return
```

Comment:
Is there a reason to call `fail` instead of `skip`? This is a parameter validation failure, not a test failure, so should be caught at the beginning by `default.bash`, if it was to be treated as a failure.

If we only catch it inside the test itself, then I think the test should be skipped and not fail.

So I think this could be just

```suggestion
    skip_k3s_version "Could not find a version of rancher-latest/rancher compatible with Kubernetes ${RD_KUBERNETES_PREV_VERSION}"
```